File: golang/divergent-modifications/good/user_repository.go
package main

type userRepository struct {
	db Database
}

func NewUserRepository(db Database) UserRepository {
	return &userRepository{db: db}
}

func (ur userRepository) UpdateUserProfile(userId int, name, email string) error {
	_, err := ur.db.Exec("UPDATE users SET name = ?, email = ? WHERE id = ?", name, email, userId)
	return err
}

func (ur userRepository) GetUserEmail(userId int) (string, error) {
	rows, err := ur.db.Query("SELECT email FROM users WHERE id = ?", userId)
	if err != nil {
		return "", err
	}
	defer rows.Close()

	if rows.Next() {
		var email string
		err := rows.Scan(&email)
		if err != nil {
			return "", err
		}
		return email, nil
	}

	return "", sql.ErrNoRows
}

File: nodejs/large-class/good/userRepository.js
const bcrypt = require('bcrypt');

class UserRepository {
    constructor(db) {
        this.db = db;
    }

    async createUser(userData) {
        // Validate user data
        this.validateUserData(userData);

        // Hash password
        userData.password = await bcrypt.hash(userData.password, 10);

        // Insert user
        const result = await this.db.query(
            'INSERT INTO users (email, password, name) VALUES ($1, $2, $3) RETURNING id',
            [userData.email, userData.password, userData.name]
        );

        return result.rows[0].id;
    }

    validateUserData(userData) {
        if (!userData.email || !userData.password || !userData.name) {
            throw new Error('Missing required fields');
        }
    }

    async getUser(userId) {
        const result = await this.db.query('SELECT * FROM users WHERE id = $1', [userId]);
        return result.rows[0];
    }
}

module.exports = UserRepository;
